dev/io.openliberty.microprofile.openapi.4.0.internal.services/src/io/openliberty/microprofile/openapi40/internal/services/impl/OpenAPIVersionConfig40Impl.java

```diff
@@ -0,0 +1,54 @@
+/*******************************************************************************
+ * Copyright (c) 2024 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package io.openliberty.microprofile.openapi40.internal.services.impl;
+
+import org.osgi.service.component.annotations.Component;
+
+import io.openliberty.microprofile.openapi20.internal.OpenAPIVersionConfigImpl;
+import io.openliberty.microprofile.openapi20.internal.services.OpenAPIVersionConfig;
+
+/**
+ * Implementation of OpenAPIVersionConfig which supports configuring version 3.0.x and 3.1.x of OpenAPI
+ */
+@Component(configurationPid = "io.openliberty.microprofile.openapi", service = OpenAPIVersionConfig.class)
+public class OpenAPIVersionConfig40Impl extends OpenAPIVersionConfigImpl {
+
+    protected static final OpenAPIVersion VERSION_31 = new OpenAPIVersion(3, 1, -1);
+    protected static final OpenAPIVersion VERSION_310 = new OpenAPIVersion(3, 1, 0);
+
+    @Override
+    protected boolean isSupported(OpenAPIVersion version) {
+        return version.getMajor() == 3
+               && (version.getMinor() == 0 || version.getMinor() == 1);
+    }
+
+    @Override
+    protected String getReplacementVersion() {
+        // If the user has configured a two-digit version, convert to the corresponding three-digit version
+        if (configuredVersion.equals(VERSION_30)) {
```

Comment:
Its a small thing but I'd suggest having a map in the superclass and calling an extra `put` in the subclass rather than copying and pasting this code.

Reply:
I'm having a think about this.

When we expect a class to be extended, it's much simpler to require the extender to implement certain methods, rather than expecting them to manipulate the state of the parent object.

Reply:
I've left the code as-is but updated the javadoc to better explain what's expected of subclasses.

I'm still a bit tempted to abstract a completely version-agnostic abstract class and hide things that subclasses shouldn't touch.